[1.19.x] Block Model Builder Root Transform Support

Rover656 opened the pull request:
Allows data generation of the root "transform" node in block model jsons (as present in `ExtendedBlockModelDeserializer`).
Use case discussed in discord: https://discord.com/channels/313125603924639766/983834532904042537/996011350990000129

Rover656 committed:
Root transform datagen

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=8860) <br/>All committers have signed the CLA.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +26,231 @@
+            transform.add("translation", writeVec3(rootTransform.translation));
+        }
+
+        if (!rootTransform.scale.equals(new Vector3f(1, 1, 1)))
```
The `new Vector3f(1, 1, 1)` should be stored in a private constant field.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +26,231 @@
+            transform.add("scale", writeVec3(rootTransform.scale));
+        }
+
+        Quaternion leftRotation = rootTransform.leftRotation;
+        if (!leftRotation.equals(Quaternion.ONE))
+        {
+            transform.add("left_rotation", writeQuaternion(leftRotation));
+        }
+
+        Quaternion rightRotation = rootTransform.rightRotation;
+        if (!rightRotation.equals(Quaternion.ONE))
+        {
+            transform.add("right_rotation", writeQuaternion(rightRotation));
+        }
```
What is more common parlance: `left_rotation` and `right_rotation`, or `rotation` and `post_rotation`? Since the deserializer code in `TransformationHelper` accepts both pairs, we should write out with the name that is more commonly understood/known.

Rover656 replied:
Yeah I think rotation and post_rotation are more descriptive.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +26,231 @@
+        private Vector3f translation = Vector3f.ZERO;
+        private Quaternion leftRotation = Quaternion.ONE.copy();
+        private Quaternion rightRotation = Quaternion.ONE.copy();
+        private Vector3f scale = new Vector3f(1, 1, 1);
```
Use the previously-suggested constant for `new Vector3f(1, 1, 1)`.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +26,231 @@
+         * @return this builder
+         * @throws NullPointerException if {@code rightRotation} is {@code null}
+         */
+        public RootTransformBuilder rightRotation(Quaternion rightRotation)
```
Should we also provide a `postRotation(Quaternion)` that calls to this method, to mimic the pair of `rotation` and `leftRotation`?

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +26,231 @@
+         */
+        public RootTransformBuilder translation(Vector3f translation)
+        {
+            Preconditions.checkNotNull(translation, "Translation must not be null");
+            this.translation = translation;
```
In general for all these methods, the `checkNotNull` can be inlined to the field assignment as it returns the value passed when not `null`.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +26,231 @@
+         * @throws NullPointerException if {@code origin} is {@code null}
+         * @throws IllegalArgumentException if {@code origin} is not {@code center}, {@code corner} or {@code opposing-corner}
+         */
+        public RootTransformBuilder origin(String origin)
+        {
+            Preconditions.checkNotNull(origin, "Origin must not be null");
+            Preconditions.checkArgument(origin.equals("center") || origin.equals("corner") || origin.equals("opposing-corner"), "Invalid value for origin name.");
```
We should be using an enum for this instead of requiring hardcoded strings. The enum should probably be `StringRepresentable`, with the enum values should include the friendly name (these three strings) and the origin vector as seen in the fields of `TransformationHelper.Deserializer`. 

Accordingly, `TransformationHelper.Deserializer#parseOrigin` would be modified to use the names from the enum values rather than hardcoded strings.

Rover656 replied:
An enum does sound like a good idea, is it a good idea to place the enum within TransformationHelper with the rest of the transformation logic, inside that common.util package or in the model package somewhere?

sciwhiz12 replied:
Having it either be a nested class of `TransformationHelper` or a nested class of `BlockModelBuilder#RootTransformBuilder` should be fine. The former would be a more logical grouping of transformation-related classes, but the latter would be better in terms of being close to where the enumeration is actually used. I'd probably say the latter, but your choice between the two.

Rover656 replied:
In 66ff28c6a3c1cfc610b99932ebe43c937e5d6089 I have added an enum to RootTransformBuilder, let me know if you'd like anything changed with it, I've implemented it as best as I could think :)

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +26,231 @@
+        return array;
+    }
+
+    public class RootTransformBuilder
+    {
```
This should have a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +26,231 @@
+    {
+        JsonObject json = super.toJson();
+
+        // Write the transform to an object
+        JsonObject transform = new JsonObject();
```
I'm thinking that the code that writes the root transform to a JSON object should be a method on `RootTransformBuilder`, which is then called from here.

Rover656 committed:
Apply the simple changes requested

Rover656 committed:
Attempt to implement the enum cleanly

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +27,282 @@
+        array.add(quaternion.k());
+        array.add(quaternion.r());
+        return array;
+    }
```
These should be moved to `RootTransformBuilder`, since that is where these methods are used.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +27,282 @@
+        {
+            Center(new Vector3f(), "origin"),
+            Corner(new Vector3f(1f, 1f, 1f), "corner"),
+            OpposingCorner(new Vector3f(.5f, .5f, .5f), "opposing_corner");
```
The convention for naming enum values is that they use `SCREAMING_SNAKE_CASE`, so these should be `CENTER`, `CORNER`, and `OPPOSING_CORNER`.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +27,282 @@
 public class BlockModelBuilder extends ModelBuilder<BlockModelBuilder> {
 
-    public BlockModelBuilder(ResourceLocation outputLocation, ExistingFileHelper existingFileHelper) {
+    private static final Vector3f ONE = new Vector3f(1, 1, 1);
```
This should be in `RootTransformBuilder`, where it is used.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +27,282 @@
+
+        private RootTransformBuilder()
+        {
+
```
Unnecessary blank space.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +27,282 @@
+            return this;
+        }
+
+        public RootTransformBuilder postRotation(Quaternion postRotation) {
```
Formatting: braces. Also, this should have javadocs.
```suggestion
        public RootTransformBuilder postRotation(Quaternion postRotation)
        {
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +27,282 @@
+
+    public class RootTransformBuilder
+    {
+        public enum Origin implements StringRepresentable
```
Should we name it a bit more particularly, like `TransformOrigin`?

Rover656 replied:
I agree, I've gone with that

sciwhiz12 replied:
Also, could you please move the enum class nearer to the bottom of the class?

Rover656 replied:
can do

Rover656 replied:
I've placed it above the private members of RootTransformBuilder to aide anyone reading this as a reference

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +27,282 @@
+            {
+                transform.addProperty("origin", origin.getSerializedName());
+            }
+            else if (originVec != null && originVec != Vector3f.ZERO)
```
Comparing against `Vector3f.ZERO` should be using `.equals(Object)`, not a reference equality check.
```suggestion
            else if (originVec != null && !originVec.equals(Vector3f.ZERO))
```

sciwhiz12 commented on src/main/java/net/minecraftforge/common/util/TransformationHelper.java:
```diff
@@ -108,9 +110,9 @@ public static boolean epsilonEquals(Vector4f v1, Vector4f v2, float epsilon)
-        private static final Vector3f ORIGIN_CORNER = new Vector3f();
-        private static final Vector3f ORIGIN_OPPOSING_CORNER = new Vector3f(1f, 1f, 1f);
-        private static final Vector3f ORIGIN_CENTER = new Vector3f(.5f, .5f, .5f);
+        private static final Vector3f ORIGIN_CORNER = Origin.Corner.vec();
+        private static final Vector3f ORIGIN_OPPOSING_CORNER = Origin.OpposingCorner.vec();
+        private static final Vector3f ORIGIN_CENTER = Origin.Center.vec();
```
It looks like we can remove these constants (which should be of no effect to binary compatibility, since these are `private`) and simply inline the three remaining uses of these constants.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -18,7 +27,282 @@
+                this.name = name;
+            }
+
+            public Vector3f vec()
```
This should be named in the style of a getter. Probably `getVector()` can suffice, unless there's a better name than just 'vector'.

Rover656 replied:
I've gone with getVector for now

Rover656 committed:
More review changes

Rover656 committed:
Move enum

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -16,9 +25,291 @@
+            }
+
+            @Override
+            public String getSerializedName()
```
Mark this method as `@NotNull` so the IDEA warning is silenced.
```suggestion
            @Override
            @NotNull
            public String getSerializedName()
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -16,9 +25,291 @@
+         */
+        public RootTransformBuilder origin(TransformOrigin origin)
+        {
+            this.origin = Preconditions.checkNotNull(origin, "Origin must not be null");;
```
Remove the unnecessary semicolon.
```suggestion
            this.origin = Preconditions.checkNotNull(origin, "Origin must not be null");
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -16,9 +25,291 @@
+        }
+
+        /**
+         * Sets the origi of the root transform.
```
Typo:
```suggestion
         * Sets the origin of the root transform.
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -16,9 +25,291 @@
+        {
+            CENTER(new Vector3f(), "origin"),
+            CORNER(new Vector3f(1f, 1f, 1f), "corner"),
+            OPPOSING_CORNER(new Vector3f(.5f, .5f, .5f), "opposing_corner");
```
It seems like these are different from the now-removed constants in `TransformationHelper`. `CENTER` should be the all-`.5f`, `CORNER` should be the all-zero, and `OPPOSING_CORNER` should be the all-one vector.

Additionally, these should be using the `Vector3f.ZERO` and `RootTransformBuilder.ONE` constants.

Rover656 replied:
Oops, I completely messed the order there, fixed :)

sciwhiz12 reviewed (changes requested):
I've poked around with this PR by trying to datagenerate the composite model test, and I've hit a pain point while writing up the code: the methods which take in a `Vector3f` should also have overloads which take in `float, float, float`, so the caller does not have to manually construct the `Vector3f` instance themselves.

As a note, I asked the opinion of @TheCurle regarding the storage of quaternion float values in JSON, if it causes a meaningful loss of precision, and they have told me that there is no such loss. 

Though it would've been neat if quaternions are serialized as degrees of rotation on the three axes, I've been advised that (converting between cartesian and polar coordinates) would cause loss of data/precision, so that's ruled out.

Also, perhaps we should provide overloads for the `Quaternion`-accepting methods which accept three values as degrees/radians of rotation for XYZay,, corresponding to `new Quaternion(x, y, z, isDegrees)`.

Rover656 committed:
Apply suggestions from code review

Co-authored-by: sciwhiz12 <sciwhiz12@gmail.com>

Rover656 committed:
Fixed the constants, completely messed the order

Rover656 committed:
Merge remote-tracking branch 'origin/root-transform-modelbuilder' into root-transform-modelbuilder

Rover656 committed:
Fix import

Rover656 committed:
Swap to jetbrains annotation

Rover656 committed:
Added more overloads

Rover656 commented:
> I've poked around with this PR by trying to datagenerate the composite model test, and I've hit a pain point while writing up the code: the methods which take in a Vector3f should also have overloads which take in float, float, float, so the caller does not have to manually construct the Vector3f instance themselves.

I have added these overloads as requested

> Also, perhaps we should provide overloads for the Quaternion-accepting methods which accept three values as degrees/radians of rotation for XYZay,, corresponding to new Quaternion(x, y, z, isDegrees).

I agree, I've gone ahead and added overloads matching the Quaternion constructor

Rover656 committed:
Fix javadoc

nikita488 commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -16,9 +25,374 @@
+
+        public enum TransformOrigin implements StringRepresentable
+        {
+            CENTER(new Vector3f(.5f, .5f, .5f), "origin"),
```
Why the name of origins was changed? Before the CENTER was "center" and OPPOSING_CORNER was "opposing-corner". Now every model that used old names inside .json files is broken in 43.1.9+

Rover656 replied:
These are the constants in code, however the third argument on the line highlighted sets the name that is used during serialisation, therefore this will not break compat. However if you are having issues with this PR, opening an issue with a bit more info might be helpful

Rover656 replied:
Ooh I have spotted the underscore issue, I'll mention it in forgecord to see if someone's free to issue a fix

nikita488 replied:
If anyone had "origin": "center" before it will parse it, but fail because now "center" is called "origin". And yes, same for underscore issue.

Rover656 replied:
Yeah just clocked that too, I'm reading on phone right now and I missed it, was going off of memory. I'll try and get this sorted out ASAP. Thanks for reporting this to me, very sorry

nikita488 replied:
All good :)
